### ISSUE TITLE
Roslyn doesn't understand pattern-matching null check for properties

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#39264]
**Version Used**: 

```
Branch master (23 Sep 2019)
Latest commit 89b4f60 by Charles Stoner:
Allow conversion of collection initializer Add extension method this arg (#38732)
```

**Steps to Reproduce**:

Compile the following code:
```cs
#nullable enable
class C
{
    string? field;

    void M1(C c)
    {
        if (c.field == null) return;
        
        c.field.ToString();
    }
    
    void M2(C c)
    {
        if (c is { field: null }) return;
        
        c.field.ToString(); // incorrect CS8602
    }
}
```
[sharplab ](https://sharplab.io/#v2:D4AQDABCCMB0AiBLAhgcwHYHsDOAXRAxtrAMKYAmApgILrIA2AntotgNwCwAUNwMToBXevWQAjepQiU64yt24gATBBLcA3twhao0MAH4IAM0SV65CAF4Ig4Zx5dtUACwQAstAAUJCAQCUm7Q0HR21EQwgPAlhjU3MLKxt6XygAdjsQ7QCMqJizWAAVTABlXAAnRHRUD1907QBfLIhGkBdXRS8ff2CtIIytMIiCCFYINSMTMwAuayF6CDrkkDTGxxXtHInyAuKyiqqaiAB6Q+H0AkxS0soCXBUigA4ANjBFRoauOqA===)

**Expected Behavior**:
Both methods report no warnings; If `c.field` is null method returns in both cases before using it.

**Actual Behavior**:
`M2` has an incorrect warning `CS8602: Dereference of a possibly null reference` reported for `c.field.ToString()`

**Notes**
It's actually even worse since such null checks makes Roslyn think that `field` is checked for null but does not update it to non-null on the false branch. Consequently Roslyn reports the warning even if `field` is declared as **non-nullable** `string`
i.e.
```
#nullable enable
class C
{
    string field;
    
    void M2(C c)
    {
        if (c is { field: null }) return;
        
        c.field.ToString(); // incorrect CS8602
    }
}
```

Note that current behavior is quite annoyting since it does not allow me to hande nulls hvia separate patterns in switches e.g.:
```cs
#nullable enable
class C
{
    string? field;
    
    void M2(C c)
    {
        switch (c)
        {
            case { field: null }: break;
            case var copy:
            {
              copy.field.ToString(); // CS8602
              c.field.ToString(); // CS8602
              break;
            }
        }
    }
}
```

**Comment by jnm2** [comment]
Is this the same issue as the one I'm running into, or should I file separately?

```cs
#nullable enable

class C
{
    void M(object? p)
    {
        switch (p)
        {
            case null:
                return;
        }

        // CS8602  Dereference of a possibly null reference.
        p.ToString();
    }
}
```

**Comment by gafter** [comment]
@jnm2 That looks like a separate issue.  Please file a new issue for that.
